### PR TITLE
Model: DiagVI (new)

### PR DESCRIPTION
**Pull request by ori-kron-wis**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -94,9 +94,11 @@ interpretability = ["captum", "shap"]
 jax = ["jax", "jaxlib", "optax", "numpyro", "flax"]
 # for custom dataloders
 dataloaders = ["lamindb>=1.12.1", "cellxgene-census", "tiledbsoma", "tiledbsoma_ml", "torchdata"]
+# for diagvi
+diagvi = ["torch_geometric", "geomloss"]
```
[the remark this answers is not in the document]
Comment (WinterHannah) [reply pyproject.toml]: Hi Can, at the moment we rely on geomloss to compute the unbalanced optimal transport loss component. However, we are working on another approach based on [POT](https://github.com/PythonOT/POT), which is more actively maintained.

**Comment by ori-kron-wis** [comment]
Hey @WinterHannah , what is the status of the PR? Should we expect more commits to it?
We can release it without docs or a tutorial, but they are very important for users to engage with the model.

**Comment by WinterHannah** [comment]
Hi @ori-kron-wis, I’ll be making a few commits to add documentation and a tutorial. It should take about a week at most!

**Review by ori-kron-wis — COMMENTED** [review]
Impressive work, everyone involved. See my several comments for review.

File: src/scvi/utils/_metrics.py
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/utils/_metrics.py:1]: We have moved the function to the diagvi utils and have updated the corresponding init.py and tutorial imports.

File: src/scvi/distributions/_lognormal.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+    logits_to_probs,
+    probs_to_logits,
+)
+
+from scvi import settings
+
+from ._constraints import optional_constraint
+
+
+class LogNormal(Distribution):
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/distributions/_lognormal.py]: We have removed the LogNormal class and ZeroInflatedLogNormal now inherits from the torch distribution.

File: src/scvi/distributions/_lognormal.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+                f"{p}: "
+                f"{self.__dict__[p] if self.__dict__[p].numel() == 1 else self.__dict__[p].size()}"
+                for p in param_names
+                if self.__dict__[p] is not None
+            ]
+        )
+        return self.__class__.__name__ + "(" + args_string + ")"
+
+
+class ZeroInflatedLogNormal(LogNormal):
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/distributions/_lognormal.py:147]: ZeroInflatedLogNormal now inherits from torch.distributions.LogNormal. We have also adapted the parameter naming due to a mismatch between scale in torch.distributions.LogNormal and our former implementation. It now follows the Normal class in _normal.py with scale denoting the standard deviation and normalized_mu the normalized mean.

File: src/scvi/distributions/_gamma.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    logits_to_probs,
+    probs_to_logits,
+)
+
+from scvi import settings
+
+from ._constraints import optional_constraint
+
+
+class Gamma(GammaTorch):
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/distributions/_gamma.py]: We have removed the Gamma class and have adjusted inheritance.

File: src/scvi/distributions/_gamma.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                f"{p}: "
+                f"{self.__dict__[p] if self.__dict__[p].numel() == 1 else self.__dict__[p].size()}"
+                for p in param_names
+                if self.__dict__[p] is not None
+            ]
+        )
+        return self.__class__.__name__ + "(" + args_string + ")"
+
+
+class ZeroInflatedGamma(Gamma):
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/distributions/_gamma.py:20]: ZeroInflatedGamma now inherits from torch.distributions.Gamma.

File: src/scvi/distributions/__init__.py
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/distributions/__init__.py:1]: I agree. We have adapted the distributions to your comments (see my answers on your comments). Is there anything else regarding the addition to the distribution folder you think should be adjusted?

File: src/scvi/external/diagvi/_model.py [hunk 6]
```diff
@@ -0,0 +1,1258 @@
...
+        return model
+
+    @torch.inference_mode()
+    def predict_celltype(
+        self,
+        labeled_modality: str,
+        target_modality: str | None = None,
+        target_adata: AnnData | None = None,
+        indices: Sequence[int] | None = None,
+        batch_size: int = 2084,
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/external/diagvi/_model.py]: Yes!

File: src/scvi/distributions/_lognormal.py [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+                f"{p}: "
+                f"{self.__dict__[p] if self.__dict__[p].numel() == 1 else self.__dict__[p].size()}"
+                for p in param_names
+                if self.__dict__[p] is not None
+            ]
+        )
+        return self.__class__.__name__ + "(" + args_string + ")"
+
+
+class Log1pNormal(Distribution):
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/distributions/_lognormal.py:20]: We now removed the LogNormal class and let ZeroInflatedLogNormal inherit from torch.distributions.LogNormal. We would therefore keep Log1pNormal as it is. Do you agree with that, or would you rather implement a more general LogNormal class with a "p" parameter?

File: CHANGELOG.md
[the remark this answers is not in the document]
Comment (michelta00) [reply CHANGELOG.md:1]: Done!

File: src/scvi/external/diagvi/_module.py [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+        # KL divergence calculation
+        if self.use_gmm_prior[mode]:
+            kl_div = inference_outputs[MODULE_KEYS.QZ_KEY].log_prob(inference_outputs["z"]).sum(
+                -1
+            ) - generative_outputs[MODULE_KEYS.PZ_KEY].log_prob(inference_outputs["z"])
+        else:
+            kl_div = kl_divergence(
+                inference_outputs[MODULE_KEYS.QZ_KEY], generative_outputs[MODULE_KEYS.PZ_KEY]
+            ).sum(dim=-1)
+        kl_local_norm = torch.sum(kl_div) / (n_obs * n_var)
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/external/diagvi/_module.py:510]: We have tested different approaches, including kl warmup as done in many other models in scvi-tools, and our current approach worked best empirically.

File: src/scvi/external/diagvi/_module.py [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+        if self.modality_likelihoods[mode] in NORMALIZE_LIKELIHOODS:
+            px_r = px_r.exp()
+        elif self.modality_likelihoods[mode] in OTHER_LIKELIHOODS and self.normalize_lib[mode]:
+            px_r = px_r.exp()
+        else:
+            px_r = F.softplus(px_r) + EPS
+
+        # Construct the appropriate generative distribution p(x) based on likelihood
+        if self.modality_likelihoods[mode] == "nb":
+            px = NegativeBinomial(px_r, logits=(px_rate + EPS).log() - px_r)
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/external/diagvi/_module.py]: You are right, thanks for catching that! We changed the NegativeBinomial initialisation and now use the (mu, theta) parameterisation correctly translating decoder outputs to distribution parameters.

File: src/scvi/external/diagvi/_model.py [hunk 10]
```diff
@@ -0,0 +1,1258 @@
...
+        accelerator, devices, device = parse_device_args(
+            accelerator=accelerator,
+            devices=devices,
+            return_device="torch",
+        )
+
+        # Extract early stopping parameters from kwargs with defaults
+        early_stopping = kwargs.pop("early_stopping", True)
+        early_stopping_monitor = kwargs.pop("early_stopping_monitor", "validation_loss")
+        early_stopping_min_delta = kwargs.pop("early_stopping_min_delta", 1)
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/external/diagvi/_model.py:234]: Yes, it is on purpose, and we have checked convergence of DiagVI with early stopping in place.

File: tests/external/diagvi/test_diagvi.py
[the remark this answers is not in the document]
Comment (michelta00) [reply tests/external/diagvi/test_diagvi.py:1]: There are tests covering predict_celltype, posterior_predictive_sample, and semi-supervised training. DiagVI currently does not support scArches or minification, but we plan to add that support in the future (which requires some adaptation due to DiagVI’s multi-adata / multi-modality structure). The same applies to DE and DA.

File: pyproject.toml [hunk 11]
```diff
@@ -95,11 +95,13 @@ interpretability = ["captum", "shap", "decoupler"]
...
```
[the remark this answers is not in the document]
Comment (michelta00) [reply pyproject.toml:99]: I think the issue is that the non-jax workflow does not install the diagvi extras since it does not install scvi-tools[tests] but scvi-tools[autotune,hub,mlflow,file_sharing,regseq,parallel,interpretability]. I have added diagvi in test_linux_nonjax.yml, which should resolve the issue.

File: docs/user_guide/models/diagvi.md
[the remark this answers is not in the document]
Comment (WinterHannah) [reply docs/user_guide/models/diagvi.md:1]: I added a note to the user guide!

File: src/scvi/distributions/__init__.py
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/distributions/__init__.py:1]: No its fine like this, and a good addition now. Hopefully, other models can use it also.

File: tests/external/diagvi/test_diagvi.py
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply tests/external/diagvi/test_diagvi.py:1]: OK. It's fine like this; we will add any capability down the road when needed.

File: docs/user_guide/models/diagvi.md
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply docs/user_guide/models/diagvi.md:1]: Thanks. This might change in the future, if we put them as mandatory dependencies

File: pyproject.toml (hunk 11 above)
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply pyproject.toml:99]: I will tend to this

File: src/scvi/distributions/_lognormal.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ori-kron-wis) [reply src/scvi/distributions/_lognormal.py:20]: It's fine, probably enough for most cases in sc.

File: src/scvi/external/diagvi/_module.py [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+
+        Returns
+        -------
+        Dictionary of inference outputs, including latent variables and
+        graph embeddings.
+        """
+        library = torch.log(x.sum(1)).unsqueeze(1)
+        graph = self.guidance_graph
+        device = x.device
+        graph = graph.to(device)
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/external/diagvi/_module.py]: We have changed the graph handling slightly, which resolves this issue.

File: src/scvi/external/diagvi/_utils.py [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    Returns
+    -------
+    Scalar tensor containing the graph reconstruction loss.
+    """
+    import torch_geometric
+
+    edge_index = graph.edge_index
+    edge_index_neg = torch_geometric.utils.structured_negative_sampling(edge_index)
+
+    pos_i = edge_index_neg[0].cpu().numpy()
```
[the remark this answers is not in the document]
Comment (michelta00) [reply src/scvi/external/diagvi/_utils.py]: Similar to above, the updated graph handling resolves this.
